index.js

```diff
@@ -0,0 +1,471 @@
+  //. | [Foldable][]    |     ✔︎     |           |           |           |
+  //. | [Traversable][] |     ✔︎     |           |           |           |
+  //. | [Apply][]       |     ✔︎     |     ✔︎     |           |           |
+  //. | [Applicative][] |           |           |           |           |
+  //. | [Chain][]       |     ✔︎     |           |           |           |
+  //. | [Monad][]       |           |           |           |           |
+  //. | [Extend][]      |           |           |           |           |
+  //. | [Comonad][]     |           |           |           |           |
+
+  //. ### Array
+
+  //# Array$concat :: Array a ~> Array a -> Array a
+  methods.Array$concat = function(xs) {
+    return this.concat(xs);
+  };
+
+  //# Array$empty :: Array a ~> Array a
+  methods.Array$empty = function() {
+    return [];
+  };
+
+  //# Array$map :: Array a ~> (a -> b) -> Array b
+  var Array$map = methods.Array$map = function(f) {
+    var result = [];
+    for (var idx = 0; idx < this.length; idx += 1) {
+      result.push(f(this[idx]));
+    }
+    return result;
+  };
+
+  //# Array$reduce :: Array a ~> ((b, a) -> b, b) -> b
+  methods.Array$reduce = function(f, x) {
+    var result = x;
+    for (var idx = 0; idx < this.length; idx += 1) {
+      result = f(result, this[idx]);
+    }
+    return result;
+  };
+
+  //  prepend :: a -> Array a -> Array a
+  var prepend = function(x) {
+    return function(xs) {
+      return [x].concat(xs);
+    };
+  };
+
+  //# Array$sequence :: Applicative f => Array (f a) ~> (a -> f a) -> f (Array a)
+  methods.Array$sequence = function(of) {
+    var applicativeXs = of([]);
+    for (var idx = this.length - 1; idx >= 0; idx -= 1) {
+      applicativeXs = ap(map(prepend, this[idx]), applicativeXs);
+    }
+    return applicativeXs;
+  };
+
+  //# Array$ap :: Array (a -> b) ~> Array a -> Array b
+  methods.Array$ap = function(xs) {
+    var result = [];
+    for (var idx = 0; idx < this.length; idx += 1) {
+      for (var idx2 = 0; idx2 < xs.length; idx2 += 1) {
+        result.push(this[idx](xs[idx2]));
+      }
+    }
+    return result;
+  };
+
+  //# Array$chain :: Array a ~> (a -> Array b) -> Array b
+  methods.Array$chain = function(f) {
+    var result = [];
+    for (var idx = 0; idx < this.length; idx += 1) {
+      push_.apply(result, f(this[idx]));
+    }
+    return result;
+  };
+
+  //. ### Function
+
+  //# Function$map :: (a -> b) ~> (b -> c) -> (a -> c)
+  methods.Function$map = function(f) {
+    var functor = this;
+    return function(x) {
+      return f(functor(x));
+    };
+  };
+
+  //# Function$ap :: (a -> b -> c) ~> (a -> b) -> (a -> c)
+  methods.Function$ap = function(f) {
+    var apply = this;
+    return function(x) {
+      return apply(x)(f(x));
+    };
+  };
+
+  //. ### Object
+
+  //# Object$concat :: StrMap a ~> StrMap a
+  methods.Object$concat = function(strMap) {
+    var result = {};
```

Comment:
This implementation is a bit hard to follow, does it really need to depend on `Array$map`? What's actually happening here?